### ISSUE TITLE
Travelling action should end on rooms bordering the target room.

### ISSUE THREAD

**Issue by karlthepagan** [ScreepsOCS/screeps.behaviour-action-pattern#626]
Difficulty: intermediate to advanced.

## The problem

The majority of pathfinding logical loops are caused by a bad approach to the target room.

### Case 1 - room geometry blocked by walls

The creep enters the target room, then finds their path is blocked by constructed walls.

### Case 2 - room geometry is less efficient than a different entrance

The creeps enters the target room, then finds their path is covered by swamps and the creep navigates outside the room in order to utilize roads.

### Case 4 (out of scope for this proposal) - determinatedSpot or determinatedTarget is occupied

When the spots are occupied creeps cecide to return to home room or path out to look for a new target. This should be resolved elsewhere.

## The Fix.

The simple fix may be to switch to moveTo with API path caching for pathfinding, but this will be less efficient in high traffic situation. Perhaps a special case when adjacent to the target room can use moveTo and a special movement mode (currently "auto" and "evade").

**Comment by karlthepagan** [comment]
Based on movement I observed in E43N81 I believe that the random? movement that happens around action.approach may lead to pathing out of the room.

Shortly after this tick: https://screeps.com/a/#!/history/E43N81?t=17129237

**Comment by karlthepagan** [comment]
Here is one observed loop:

```
{"creepName":"reserver-gammaRMine-2","pos":{"x":49,"y":17,"roomName":"E47N81"},"leaveBorder":7,"nopath":"zero len","drive":"nopath","Creep":"drive"}
{"creepName":"reserver-gammaRMine-2","run":"reserving","claimer":"run","Behaviour":"claimer"}

{"creepName":"reserver-gammaRMine-2","pos":{"x":48,"y":17,"roomName":"E47N81"},"Behaviour":"claimer","Creep":"run"}
{"creepName":"reserver-gammaRMine-2","assign":"reserving","target":"gammaRMine","Action":"assign"}
{"creepName":"reserver-gammaRMine-2","nextAction":"reserving","reserve":"nextAction","Task":"reserve"}
{"creepName":"reserver-gammaRMine-2","pos":{"x":48,"y":17,"roomName":"E47N81"},"direction":3,"moveResult":0,"drive":"ok","Creep":"drive"}
{"creepName":"reserver-gammaRMine-2","run":"reserving","claimer":"run","Behaviour":"claimer"}

{"creepName":"reserver-gammaRMine-2","pos":{"x":0,"y":17,"roomName":"E48N81"},"Behaviour":"claimer","Creep":"run"}
{"creepName":"reserver-gammaRMine-2","pos":{"x":0,"y":17,"roomName":"E48N81"},"direction":4,"moveResult":0,"drive":"ok","Creep":"drive"}
{"creepName":"reserver-gammaRMine-2","run":"reserving","claimer":"run","Behaviour":"claimer"}

{"creepName":"reserver-gammaRMine-2","pos":{"x":1,"y":18,"roomName":"E48N81"},"Behaviour":"claimer","Creep":"run"}
{"creepName":"reserver-gammaRMine-2","assign":"reserving","target":"5873bd9011e3e4361b4d96e1","Action":"assign"}
{"creepName":"reserver-gammaRMine-2","nextAction":"reserving","reserve":"nextAction","Task":"reserve"}
{"creepName":"reserver-gammaRMine-2","pos":{"x":1,"y":18,"roomName":"E48N81"},"direction":8,"moveResult":0,"drive":"ok","Creep":"drive"}
{"creepName":"reserver-gammaRMine-2","run":"reserving","claimer":"run","Behaviour":"claimer"}

{"creepName":"reserver-gammaRMine-2","pos":{"x":49,"y":17,"roomName":"E47N81"},"Behaviour":"claimer","Creep":"run"}
{"creepName":"reserver-gammaRMine-2","pos":{"x":49,"y":17,"roomName":"E47N81"},"leaveBorder":7,"nopath":"zero len","drive":"nopath","Creep":"drive"}
{"creepName":"reserver-gammaRMine-2","run":"reserving","claimer":"run","Behaviour":"claimer"}

{"creepName":"reserver-gammaRMine-2","pos":{"x":48,"y":17,"roomName":"E47N81"},"Behaviour":"claimer","Creep":"run"}
{"creepName":"reserver-gammaRMine-2","assign":"reserving","target":"gammaRMine","Action":"assign"}
{"creepName":"reserver-gammaRMine-2","nextAction":"reserving","reserve":"nextAction","Task":"reserve"}
{"creepName":"reserver-gammaRMine-2","pos":{"x":48,"y":17,"roomName":"E47N81"},"direction":3,"moveResult":0,"drive":"ok","Creep":"drive"}
```

my current conclusion is that this is more fundamental to pathfinding and not the direct fault of the travelling action.

**Comment by karlthepagan** [comment]
It would be worth examining adopting the bonzAI pathing package: https://github.com/bonzaiferroni/bonzAI/wiki/Traveler-API
